### PR TITLE
Add Workflows

### PR DESCRIPTION
**Pull request by AlaaElattar**

### CODE REVIEW

**Comment by AlaaElattar** [comment]
- According to the unknown value in the codecov badge will be updated after merging with main.

**Review by AhmedHanafy725 — COMMENTED** [review]
please resolve the conflicts here

**Comment by AhmedHanafy725** [comment]
do we need to keep the coverage directory uploaded?

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+name: Build
+
+on:
+  pull_request:
```
Comment (AhmedHanafy725) [inline .github/workflows/build.yaml:4]: please add on push on the main branch too.

File: .github/workflows/build.yaml
Comment (AhmedHanafy725) [inline .github/workflows/build.yaml:1]: the file name is called `build` while it doesn't build the code. maybe call it `analyze`

File: .github/workflows/build.yaml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        uses: dart-lang/setup-dart@v1.6.0
+        with:
+          sdk: "stable"
+      - name: Install Melos
+        run: dart pub global activate melos
+
+      - name: Install dependencies
+        run: dart pub global run melos bootstrap
+
+      - name: Linting
```
Comment (AhmedHanafy725) [inline .github/workflows/build.yaml:26]:
```suggestion
      - name: Analyze
```

File: .github/workflows/coverage.yaml [hunk 3]
```diff
@@ -0,0 +1,51 @@
+name: Upload Coverage
+
+on:
+  pull_request:
+    branches:
+      - main
```
Comment (AhmedHanafy725) [inline .github/workflows/coverage.yaml:6]: same here add on pushing on the main branch.

File: .github/workflows/coverage.yaml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        run: melos run unit_test
+
+      - name: Create coverage directory
+        run: mkdir -p coverage
+
+      - name: Merge coverage
+        run: |
+          melos exec rm -rf coverage
+          melos run unit_test
+          melos exec --dir-exists="coverage" -- bash -c 'dart pub global run coverde filter --input coverage/lcov.info --output coverage/filtered.lcov.info --filters .g.dart'
```
Comment (AhmedHanafy725) [inline .github/workflows/coverage.yaml:44]: please split running the tests and calculating the coverage into 2 different steps. the coverage step should check if the tests passed successfully to run.

File: .github/workflows/coverage.yaml [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+          melos exec --dir-exists="coverage" -- bash -c 'dart pub global run coverde value -i coverage/filtered.lcov.info > coverage/result.txt'
+ 
+      - name: Upload coverage to Codecov
+        uses: codecov/codecov-action@v3
```
Comment (AhmedHanafy725) [inline .github/workflows/coverage.yaml:48]: this step should check if the coverage step is finished without errors

File: .github/workflows/tests.yaml
Comment (AhmedHanafy725) [inline .github/workflows/tests.yaml:1]: this workflow is a duplicate of the coverage workflow. let's remove it

**Review by AhmedHanafy725 — CHANGES_REQUESTED** [review]

**Review by AhmedHanafy725 — APPROVED** [review]

**Pull request merged**